### PR TITLE
Move from untyped->typed config

### PR DESCRIPTION
**Pull request by lemonmade**
I'm working through adding a linting command for Sewing Kit, and it's much easier if the "base" config of the project has no type-checking commands; it lets us automatically create a "typed" version for CI (this doesn't work well right now because the editor config needs to extend the "full" config, then disable type checking, which means the "full" config always needs to exist).

### CODE REVIEW

**Comment by GoodForOneFare** [comment]
Is the suggestion that:
1. There will be less friction when bootstrapping a project because only one config is necessary
2. CI will automatically do a fully-typed lint, even if the user has not defined a script to run tslint with `--type-check`

?

**Comment by lemonmade** [comment]
The way I have this working is that, when you run the sewing-kit lint command, it creates a private config file that extends the project's config (which now just includes the default/ React configs) and uses the correct CLI arguments to get type checking. This means only one config file for the project, which makes things a lot simpler

**Comment by ismail-syed** [comment]
I think we'll need to change [`./build/untyped`](https://github.com/Shopify/tslint-config-shopify/blob/typed-config/tslint.json#L3) to ~~`./build/typed`~~ `./build/base` in this project's `tslint.json`

File: README.md [hunk 1]
```diff
@@ -68,35 +68,33 @@ Override these rules in `tslint.json`:
...
 ```
 
-Then update your CLI tslint script to run against `tslint-full.json`.
+And, in your `lint` script, make sure to run it with the appropriate command-line arguments to get the full benefit of this configuration:
 
 ```json
 {
   "scripts": {
-    "tslint": "tslint -c ./config/tslint-full.json './src/**/*.{ts,tsx}' --project tsconfig.json --type-check"
+    "tslint": "tslint --config ./config/tslint/full.json './src/**/*.{ts,tsx}' --project tsconfig.json --type-check"
```
Comment (ismail-syed) [inline README.md]: From the [`v5.4.0`](https://github.com/palantir/tslint/releases/tag/5.4.0) release: 

>The --type-check CLI option now only enables checking for compiler errors before linting. It is no longer required to enable rules that use the type checker -- only need to supply --project now.

Forgot to update the docs when I last updated last time. 😞 
If you could remove the `--type-check` in here, that would be 👍

**Review by ismail-syed — CHANGES_REQUESTED** [review]
Giving this a 🎩 through Neutron atm.

I like this approach better, it simplifies the onboarding process for consumers. 

However, previously every rule in the `untyped` config would be turned off, and whether we like that rule to be `on/off` is based on it's setting in `typescript.ts`, `functionality.ts` etc.  

We have a bit off the opposite now. I propose that we keep only one setting for the type-checking rules (which will live in our `typed.ts` config) and remove all the one we have turned off in `typescript.ts`, `functionality.ts` etc.

File: README.md (hunk 1 above)
Comment (ismail-syed) [reply README.md]: I also noticed a 1s increase in performance within this repo when we ran `tslint`. If we update subsequent repo's we should remove `--type-check`, I think we'll see some a small performance boost.

**Comment by ismail-syed** [comment]
Finished 🎩 with Neutron. No issues 👍

**Review by ismail-syed — APPROVED** [review]
💯 

🚢

**Comment by GoodForOneFare** [comment]
Please don't ship yet.

**Comment by ismail-syed** [comment]
Gord says 🙅‍♂️

**Review by GoodForOneFare — COMMENTED** [review]
What's the benefit of the untyped + full config nowadays?

In the past, the typed config made VS Code unusable.  Now, a few `foo-bar-rule requires type checking` terminal messages appear on startup, but the plugin still functions beautifully.

Atom's tslint plugin works without any problems, too.

I think this PR should either:
1. Do away with the two-config concept entirely

2. If we're still worried about the `requires type checking` warnings, we shouldn't ship this because those messages are easily introduced by project-specific rule overrides

**Comment by ismail-syed** [comment]
Thanks for remembering to 🎩  with Atom!
❤️ 

Atom's error messages a super obnoxious so it's easy to see if things break.

**Comment by ismail-syed** [comment]
The config that editors pick up shouldn't have any type checking rules enabled, we shouldn't be seeing `foo-bar-rule requires type checking`. 🤔 

Let's sit down together tomorrow and see where this 💩 the bed.

**Comment by GoodForOneFare** [comment]
There's no 💩, AFAIC.  I'm okay with VS Code displaying some warnings buried in a console.

I think the config split is an artifact of when the lint plugins gave ridiculous feedback on typed rules (popped up a warning dialog on every save, IIRC?)

-----

You can easily get type warnings by adding an override.  e.g.,

Create `custom-rules.js`:
```
module.exports = {
  rules: {
    "await-promise": true,
  },
};
```

Then adding them after the base rules:
```
{
  "extends": [
    "tslint-config-shopify",
    "./custom-rules.js"
  ]
}
```

**Comment by ismail-syed** [comment]
Was able to reproduce the `Warning: The 'await-promise' rule requires type checking` in the VSCode output console tab. 

I'm okay with it buried in the console a well. 
But I'm not okay with a pop-up warning on every save. I'm not seeing any of those. @GoodForOneFare have you noticed any of those obnoxious pop-up warnings with this new setup.

**Comment by lemonmade** [comment]
Based on these discussions, I completely wiped out my changes and got rid of the multiple config stuff entirely. Can you guys take another look?

**Comment by ismail-syed** [comment]
LGTM 👍 

> What's the benefit of the untyped + full config nowadays?
> In the past, the typed config made VS Code unusable. 

Were there recent changes to VSCode that allowed this work?

**Comment by ismail-syed** [comment]
Maye this? 
https://code.visualstudio.com/updates/v1_11#_source-control

**Comment by GoodForOneFare** [comment]
No, not that.  tbh, I can't even remember which editor was problematic.  I was more of an Atom user back then, so you could start by searching through the `linter-tslint` repo for answer.

**Comment by GoodForOneFare** [comment]
Out of paranoia, I also tested in Sublime.  Displays the same one-time console errors as vscode, and works perfectly.

connection ts shrink-ray 2017-06-27 13-59-54

**Review by GoodForOneFare — APPROVED** [review]
LGTM 👍 Tried this out in Atom/VSCode (using shrink-ray, cuz neutron master is busted right now).

(I accidentally approved this in the _sewing-kit next_ PR 😕 )

**Comment by GoodForOneFare** [comment]
While reviewing the neutron tslint bump, I found [this post](https://github.com/Microsoft/vscode-tslint/issues/70#issuecomment-290744318) about how VS Code used to do _nothing_ when typed rules were present:

> This behavior was changed in TSLint recently (palantir/tslint#2169) so that TSLint now just prints a warning message and continues linting. That should remove one of the big issues mentioned above, so it ought to be possible to just use one tslint.json file now instead of multiple.

So that explains the convoluted nature of `tslint-config-shopify@<3`.

**Pull request merged**